README.md
# Format-Mail ✉️

Formatter for email, making your email look beautiful! 😍

## ℹ️ Info

Format-Mail is a simple Node.js module for formatting email contents. With this module, you can add images, headings, and styles to your email to make it visually appealing. It is specifically designed for use with [Nodemailer](https://npmjs.com/nodemailer), a popular email sending library.

## Installation

```sh
npm install format-mail@latest
```

Or, if you prefer using Yarn:

```sh
yarn add format-mail@latest
```

## Basic Formatting Example (JavaScript)

```javascript
'use strict';
const nodemailer = require('nodemailer');
const { emailFormatter, HTMLTags } = require('format-mail');

const transporter = nodemailer.createTransport({
  host: 'smtp.forwardemail.net',
  port: 465,
  secure: true,
  auth: {
    // TODO: Replace `user` and `pass` values from <https://forwardemail.net>
    user: 'REPLACE-WITH-YOUR-ALIAS@YOURDOMAIN.COM',
    pass: 'REPLACE-WITH-YOUR-GENERATED-PASSWORD',
  },
});

// async..await is not allowed in the global scope, so we wrap the code in an async function
async function main() {
  // Send mail with the defined transport object
  const info = await transporter.sendMail({
    from: '"Fred Foo 👻" <foo@example.com>', // Sender address
    to: 'bar@example.com, baz@example.com', // List of receivers
    subject: 'Hello ✔', // Subject line
    text: 'Hello world?', // Plain text body
    html: `${emailFormatter.bold('Hello world?')}\n${HTMLTags.heading1(
      'Hello world!',
      'text-align: center',
      'display: flex',
      'justify-contents: center',
      'align-items: center'
    )}`, // HTML body with formatted text and basic styles
  });

  console.log('Message sent: %s', info.messageId);
  // Output: Message sent: <b658f8ca-6296-ccf4-8306-87d57a0b4321@example.com>

  //
  // NOTE: You can go to https://forwardemail.net/my-account/emails to see the delivery status and preview of your email.
  //       Alternatively, you can use the "preview-email" npm package to preview emails locally in browsers and iOS Simulator.
  //       Check out the repository at <https://github.com/forwardemail/preview-email>.
  //
}

main().catch(console.error);
```

## Basic Formatting Example (TypeScript)

```typescript
import * as nodemailer from 'nodemailer';
import { emailFormatter, HTMLTags } from 'format-mail';

const transporter = nodemailer.createTransport({
  host: 'smtp.forwardemail.net',
  port: 465,
  secure: true,
  auth: {
    // TODO: Replace `user` and `pass` values from <https://forwardemail.net>
    user: 'REPLACE-WITH-YOUR-ALIAS@YOURDOMAIN.COM',
    pass: 'REPLACE-WITH-YOUR-GENERATED-PASSWORD',
  },
});

async function main() {
  const info = await transporter.sendMail({
    from: '"Fred Foo 👻" <foo@example.com>', // Sender address
    to: 'bar@example.com, baz@example.com', // List of receivers
    subject: 'Hello ✔', // Subject line
    text: 'Hello world?', // Plain text body
    html: `${emailFormatter.bold('Hello world?')}\

n${HTMLTags.heading1(
      'Hello world!',
      'text-align: center',
      'display: flex',
      'justify-contents: center',
      'align-items: center'
    )}`, // HTML body with formatted text and basic styles
  });

  console.log('Message sent: %s', info.messageId);
  // Output: Message sent: <b658f8ca-6296-ccf4-8306-87d57a0b4321@example.com>

  //
  // NOTE: You can go to https://forwardemail.net/my-account/emails to see the delivery status and preview of your email.
  //       Alternatively, you can use the "preview-email" npm package to preview emails locally in browsers and iOS Simulator.
  //       Check out the repository at <https://github.com/forwardemail/preview-email>.
  //
}

main().catch(console.error);
```

Feel free to explore the Format-Mail module and unleash your creativity in formatting beautiful emails! 😊🎉


changelog.md
# Changelog

## 1.0.0 - 2023-06-25

### Added

- `bold()` method
- `italic()` method
- `img()` method
- `font()` method
- `size()` method
- `strikethrough()` method
- `color()` method
- `heading1()` method
- `heading2()` method
- `heading3()` method
- `link()` method

## 1.1.0 - 2023-06-26

### Added

- `paragraph()` method
- `div()` method

### Fixed

- Fixed markdown typos

## 1.2.0 - 2023-06-23

### Changed

- Font Enum

### Added

- `table()` method
- `td()` method
- `th()` method
- `tr()` method
- `br()` method
- `hr()` method
- `button()` method
- `span()` method
- `ol()` method
- `ul()` method

### Fixed

- Fixed error handler


test/module.test.js
const { emailFormatter, Fonts, Align } = require('../dist');

describe('emailFormatter', () => {
  test('formatting with bold', () => {
    const formattedText = emailFormatter.bold('Hello');
    expect(formattedText).toBe('<strong>Hello</strong>');
  });

  test('formatting with font', () => {
    const fontName = Fonts.Abel;
    const formattedText = emailFormatter.font(fontName, 'Hello');
    expect(formattedText).toBe('<span style="font-family: Abel">Hello</span>');
  });

  test('formatting with italic', () => {
    const formattedText = emailFormatter.italic('Hello');
    expect(formattedText).toBe('<em>Hello</em>');
  });

  test('formatting with color', () => {
    const formattedText = emailFormatter.color(
      emailFormatter.hex('FFFFFF'),
      'Hello!'
    );
    expect(formattedText).toBe('<span style="color: #FFFFFF">Hello!</span>');
  });
  test('formatting with color2', () => {
    const formattedText = emailFormatter.color(
      emailFormatter.rgba(255, 0, 0, 0.4),
      'Hello!'
    );
    expect(formattedText).toBe(
      '<span style="color: rgba(255, 0, 0, 0.4)">Hello!</span>'
    );
  });
  test('formatting with color3', () => {
    const formattedText = emailFormatter.color(
      emailFormatter.rgb(255, 0, 0),
      'Hello!'
    );
    expect(formattedText).toBe(
      '<span style="color: rgb(255, 0, 0)">Hello!</span>'
    );
  });

  test('formatting with image', () => {
    const formattedImage = emailFormatter.img(
      'https://upload.wikimedia.org/wikipedia/commons/thumb/4/4c/Typescript_logo_2020.svg/1200px-Typescript_logo_2020.svg.png'
    );
    expect(formattedImage).toBe(
      '<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/4/4c/Typescript_logo_2020.svg/1200px-Typescript_logo_2020.svg.png" />'
    );
  });

  test('formatting with strikethrough', () => {
    const formattedText = emailFormatter.strikethrough('Hello!');

    expect(formattedText).toBe('<s>Hello!</s>');
  });

  test('formatting with underline', () => {
    const formattedText = emailFormatter.underline('Hello!');

    expect(formattedText).toBe('<u>Hello!</u>');
  });

  test('formatting with size', () => {
    const newSizeText = emailFormatter.size('1rem', 'Hello');

    expect(newSizeText).toBe('<span style="font-size: 1rem">Hello</span>');
  });
  test('Align a string', () => {
    const AlignedText = emailFormatter.align(Align.center, 'Hello World!');
    expect(AlignedText).toBe(
      `<span style="text-align: center">Hello World!</span>`
    );
  });
});


source/email/htmlTags.ts
/**
 * Returns the type of the given identifier.
 * @param {*} identifier - The identifier to check the type of.
 * @returns {number | "string" | "number" | "bigint" | "boolean" | "symbol" | "undefined" | "object" | "function" | null} - The type of the identifier.
 */

const getType = (identifier: any) => {
  if (identifier === Number.isNaN) {
    return NaN;
  } else if (identifier === null) {
    return null;
  } else {
    return typeof identifier;
  }
};

/**
 * Represents the type of error.
 * @typedef {("Error"|"TypeError"|"SyntaxError"|"ReferenceError"|"RangeError"|"Custom")} ErrorType
 */

type ErrorType =
  | 'Error'
  | 'TypeError'
  | 'SyntaxError'
  | 'ReferenceError'
  | 'RangeError'
  | 'Custom';

/**
 * Creates a custom error class with the given error name.
 * @param {string} errorName - The name of the custom error class.
 * @returns {any} - The created custom error class.
 */
function createCustomErrorClass(errorName: string) {
  return eval(`
      class ${errorName} extends Error {
        constructor(message) {
          const errorMessage = message.join(" ");
          super(\`${errorName}\${errorMessage}\`);
          this.name = "${errorName}";
          Object.setPrototypeOf(this, new.target.prototype);
        }
      }
      
      ${errorName};
    `);
}

/**
 * Throws an error with the specified error type, error name, and message.
 * @param {ErrorType} errorType - The type of the error.
 * @param {string} errorName - The name of the error.
 * @param {...string} message - The error message.
 * @returns {void}
 */
function throwError(
  errorType: ErrorType,
  errorName: string,
  ...message: string[]
) {
  if (errorType === 'Custom') {
    const customErrorClass = createCustomErrorClass(errorName);
    throw new customErrorClass(message);
  }

  const errorConstructor = globalThis[errorType] || Error;
  const errorMessage = message.join('\n');
  throw new errorConstructor(`${errorName}\n${errorMessage}`);
}

/**
 * Object with the HTML Tags.
 * @namespace
 */
export const HTMLTags = {
  /**
   * Generates an h1 tag with optional styles.
   * @param {string} str - The content of the h1 tag.
   * @param {...string} style - Optional styles to be applied.
   * @returns {string} The generated h1 tag HTML.
   */
  heading1(str: string, ...style: string[]) {
    if (typeof str !== 'string') {
      throwError(
        'Custom',
        'InvalidHeadingType',
        `Expected type of 'STR' to be 'string' but got ${getType(str)} instead.`
      );
    }
    const styleString =
      style.length > 0 ? style.map((s: string) => s + ';').join(' ') : '';
    return `<h1${
      styleString ? ' style="' + styleString + '"' : ''
    }>${str}</h1>`;
  },

  /**
   * Generates an h2 tag with optional styles.
   * @param {string} str - The content of the h2 tag.
   * @param {...string} style - Optional styles to be applied.
   * @returns {string} The generated h2 tag HTML.
   */
  heading2(str: string, ...style: string[]) {
    if (typeof str !== 'string') {
      throwError(
        'Custom',
        'InvalidHeadingType',
        `Expected type of 'STR' to be 'string' but got ${getType(str)} instead.`
      );
    }
    const styleString =
      style.length > 0 ? style.map((s: string) => s + ';').join(' ') : '';
    return `<h2${
      styleString ? ' style="' + styleString + '"' : ''
    }>${str}</h2>`;
  },

  /**
   * Generates an h3 tag with optional styles.
   * @param {string} str - The content of the h3 tag.
   * @param {...string} style - Optional styles to be applied.
   * @returns {string} The generated h3 tag HTML.
   */
  heading3(str: string, ...style: string[]) {
    if (typeof str !== 'string') {
      throwError(
        'Custom',
        'InvalidHeadingType',
        `Expected type of 'STR' to be 'string' but got ${getType(str)} instead.`
      );
    }
    const styleString =
      style.length > 0 ? style.map((s: string) => s + ';').join(' ') : '';
    return `<h3${
      styleString ? ' style="' + styleString + '"' : ''
    }>${str}</h3>`;
  },

  /**
   * Generates an anchor tag with optional styles.
   * @param {string} href - The URL for the anchor tag.
   * @param {string} str - The text content of the anchor tag.
   * @param {...string} style - Optional styles to be applied.
   * @returns {string} The generated anchor tag HTML.
   */
  link(href: string, str: string, ...style: string[]) {
    if (typeof href !== 'string') {
      throwError(
        'Custom',
        'InvalidHeadingType',
        `Expected type of 'href' to be 'string' but got ${getType(
          href
        )} instead.`
      );
    }
    if (typeof str !== 'string') {
      throwError(
        'Custom',
        'InvalidHeadingType',
        `Expected type of 'STR' to be 'string' but got ${getType(str)} instead.`
      );
    }
    const styleString =
      style.length > 0 ? style.map((s: string) => s + ';').join(' ') : '';
    return `<a href=${href}${
      styleString ? ' style="' + styleString + '"' : ''
    }>${str}</a>`;
  },
  /**
   * Formats the given URL into an image tag.
   * @param {string} url - The URL of the image.
   * @param {...string} style - Optional styles to be applied.
   * @returns {string} - The formatted image tag.
   */
  img(url: string, ...style: string[]) {
    if (typeof url !== 'string') {
      throwError(
        'Custom',
        'EmailFormatterError',
        `Expected type of 'URL' to be 'string' but received ${getType(url)}`
      );
    }
    const styleString =
      style.length > 0 ? style.map((s: string) => s + ';').join(' ') : '';
    return `<img src="${url}"${
      styleString ? ' style="' + styleString + '"' : ''
    }/>`;
  },
  /**
   * Generates a paragraph tag with optional styles.
   * @param {string} str - The content of the h3 tag.
   * @param {...string} style - Optional styles to be applied.
   * @returns {string} The generated paragraph tag HTML.
   */
  paragraph(str: string, ...style: string[]) {
    if (typeof str !== 'string') {
      throwError(
        'Custom',
        'InvalidParagraphType',
        `Expected type of 'STR' to be 'string' but got ${getType(str)} instead.`
      );
    }
    const styleString =
      style.length > 0 ? style.map((s: string) => s + ';').join(' ') : '';
    return `<p${styleString ? ' style="' + styleString + '"' : ''}>${str}</p>`;
  },
  /**
   * Generates a div tag with optional styles.
   * @param {string} str - The content of the h3 tag.
   * @param {...string} style - Optional styles to be applied.
   * @returns {string} The generated div tag HTML.
   */
  div(str: string, ...style: string[]) {
    if (typeof str !== 'string') {
      throwError(
        'Custom',
        'InvalidDivType',
        `Expected type of 'STR' to be 'string' but got ${getType(str)} instead.`
      );
    }
    const styleString =
      style.length > 0 ? style.map((s: string) => s + ';').join(' ') : '';
    return `<div${
      styleString ? ' style="' + styleString + '"' : ''
    }>${str}</div>`;
  },
  /**
   * Generates a span tag with optional styles.
   * @param {string} str - The content of the span tag.
   * @param {...string} style - Optional styles to be applied.
   * @returns {string} The generated span tag HTML.
   */
  span(str: string, ...style: string[]) {
    if (typeof str !== 'string') {
      throwError(
        'Custom',
        'InvalidSpanType',
        `Expected type of 'STR' to be 'string' but got ${getType(str)} instead.`
      );
    }
    const styleString =
      style.length > 0 ? style.map((s: string) => s + ';').join(' ') : '';
    return `<span${
      styleString ? ' style="' + styleString + '"' : ''
    }>${str}</span>`;
  },

  /**
   * Generates an unordered list (ul) tag with optional styles.
   * @param {string[]} items - The array of list items.
   * @param {...string} style - Optional styles to be applied.
   * @returns {string} The generated ul tag HTML.
   */
  ul(items: string[], ...style: string[]) {
    const itemTags = items.map((item) => `<li>${item}</li>`).join('');
    const styleString =
      style.length > 0 ? style.map((s: string) => s + ';').join(' ') : '';
    return `<ul${
      styleString ? ' style="' + styleString + '"' : ''
    }>${itemTags}</ul>`;
  },

  /**
   * Generates an ordered list (ol) tag with optional styles.
   * @param {string[]} items - The array of list items.
   * @param {...string} style - Optional styles to be applied.
   * @returns {string} The generated ol tag HTML.
   */
  ol(items: string[], ...style: string[]) {
    const itemTags = items.map((item) => `<li>${item}</li>`).join('');
    const styleString =
      style.length > 0 ? style.map((s: string) => s + ';').join(' ') : '';
    return `<ol${
      styleString ? ' style="' + styleString + '"' : ''
    }>${itemTags}</ol>`;
  },

  /**
   * Generates a horizontal rule (hr) tag with optional styles.
   * @param {...string} style - Optional styles to be applied.
   * @returns {string} The generated hr tag HTML.
   */
  hr(...style: string[]) {
    const styleString =
      style.length > 0 ? style.map((s: string) => s + ';').join(' ') : '';
    return `<hr${styleString ? ' style="' + styleString + '"' : ''} />`;
  },

  /**
   * Generates a line break (br) tag.
   * @returns {string} The generated br tag HTML.
   */
  br() {
    return `<br />`;
  },

  /**
   * Generates a table (table) tag with optional styles.
   * @param {string} content - The content of the table.
   * @param {...string} style - Optional styles to be applied.
   * @returns {string} The generated table tag HTML.
   */
  table(content: string, ...style: string[]) {
    if (typeof content !== 'string') {
      throwError(
        'Custom',
        'InvalidParagraphType',
        `Expected type of 'STR' to be 'string' but got ${getType(
          content
        )} instead.`
      );
    }

    const styleString =
      style.length > 0 ? style.map((s: string) => s + ';').join(' ') : '';
    return `<table${
      styleString ? ' style="' + styleString + '"' : ''
    }>${content}</table>`;
  },

  /**
   * Generates a table row (tr) tag with optional styles.
   * @param {string} content - The content of the table row.
   * @param {...string} style - Optional styles to be applied.
   * @returns {string} The generated tr tag HTML.
   */
  tr(content: string, ...style: string[]) {
    if (typeof content !== 'string') {
      throwError(
        'Custom',
        'InvalidParagraphType',
        `Expected type of 'STR' to be 'string' but got ${getType(
          content
        )} instead.`
      );
    }

    const styleString =
      style.length > 0 ? style.map((s: string) => s + ';').join(' ') : '';
    return `<tr${
      styleString ? ' style="' + styleString + '"' : ''
    }>${content}</tr>`;
  },

  /**
   * Generates a table data (td) tag with optional styles.
   * @param {string} content - The content of the table data.
   * @param {...string} style - Optional styles to be applied.
   * @returns {string} The generated td tag HTML.
   */
  td(content: string, ...style: string[]) {
    if (typeof content !== 'string') {
      throwError(
        'Custom',
        'InvalidParagraphType',
        `Expected type of 'STR' to be 'string' but got ${getType(
          content
        )} instead.`
      );
    }

    const styleString =
      style.length > 0 ? style.map((s: string) => s + ';').join(' ') : '';
    return `<td${
      styleString ? ' style="' + styleString + '"' : ''
    }>${content}</td>`;
  },

  /**
   * Generates a table header (th) tag with optional styles.
   * @param {string} content - The content of the table header.
   * @param {...string} style - Optional styles to be applied.
   * @returns {string} The generated th tag HTML.
   */
  th(content: string, ...style: string[]) {
    if (typeof content !== 'string') {
      throwError(
        'Custom',
        'InvalidParagraphType',
        `Expected type of 'STR' to be 'string' but got ${getType(
          content
        )} instead.`
      );
    }

    const styleString =
      style.length > 0 ? style.map((s: string) => s + ';').join(' ') : '';
    return `<th${
      styleString ? ' style="' + styleString + '"' : ''
    }>${content}</th>`;
  },

  /**
   * Generates a button tag with optional styles and onClick event handler.
   * @param {string} text - The text content of the button.
   * @param {Object} options - The options for the button.
   * @param {Function} options.onClick - The onClick event handler function.
   * @param {...string} style - Optional styles to be applied.
   * @returns {string} The generated button tag HTML.
   */
  button(text: string, options: { onClick: Function }, ...style: string[]) {
    const onClickHandler = options.onClick
      ? `onclick="${options.onClick.toString()}"`
      : '';

    const styleString =
      style.length > 0 ? style.map((s: string) => s + ';').join(' ') : '';

    return `<button${onClickHandler}${
      styleString ? ' style="' + styleString + '"' : ''
    }>${text}</button>`;
  },
};


source/email/formatter.ts
import { Align, Fonts } from '../enums';

/**
 * Returns the type of the given identifier.
 * @param {*} identifier - The identifier to check the type of.
 * @returns {number | "string" | "number" | "bigint" | "boolean" | "symbol" | "undefined" | "object" | "function" | null} - The type of the identifier.
 */

const getType = (identifier: any) => {
  if (identifier === Number.isNaN) {
    return NaN;
  } else if (identifier === null) {
    return null;
  } else {
    return typeof identifier;
  }
};

/**
 * Represents the type of error.
 * @typedef {("Error"|"TypeError"|"SyntaxError"|"ReferenceError"|"RangeError"|"Custom")} ErrorType
 */

type ErrorType =
  | 'Error'
  | 'TypeError'
  | 'SyntaxError'
  | 'ReferenceError'
  | 'RangeError'
  | 'Custom';

/**
 * Creates a custom error class with the given error name.
 * @param {string} errorName - The name of the custom error class.
 * @returns {any} - The created custom error class.
 */
function createCustomErrorClass(errorName: string) {
  return eval(`
    class ${errorName} extends Error {
      constructor(message) {
        const errorMessage = message.join(" ");
        super(\`${errorName}\\n\${errorMessage}\`);
        this.name = "${errorName}";
        Object.setPrototypeOf(this, new.target.prototype);
      }
    }
    
    ${errorName};
  `);
}

/**
 * Throws an error with the specified error type, error name, and message.
 * @param {ErrorType} errorType - The type of the error.
 * @param {string} errorName - The name of the error.
 * @param {...string} message - The error message.
 * @returns {void}
 */
function throwError(
  errorType: ErrorType,
  errorName: string,
  ...message: string[]
) {
  if (errorType === 'Custom') {
    const customErrorClass = createCustomErrorClass(errorName);
    throw new customErrorClass(message);
  }

  const errorConstructor = globalThis[errorType] || Error;
  const errorMessage = message.join('\n');
  throw new errorConstructor(`${errorName}\n${errorMessage}`);
}

/**
 * The emailFormatter object with various formatting functions.
 * @namespace
 */
export const emailFormatter = {
  /**
   * Formats the given string into bold text.
   * @param {string} str - The text you want to be bold.
   * @returns {string} - The formatted bold text.
   */
  bold(text?: string) {
    const str = text ?? 'Hello, World!';
    if (typeof str !== 'string') {
      throwError(
        'Custom',
        'EmailFormatterError',
        `Expected type of 'STR' to be 'string' but received ${getType(str)}`
      );
    }
    return `<strong>${str}</strong>`;
  },

  /**
   * Formats the given string into italic text.
   * @param {string} str - The text you want to be italic.
   * @returns {string} - The formatted italic text.
   */
  italic(text?: string) {
    const str = text ?? 'Hello, World!';

    if (typeof str !== 'string') {
      throwError(
        'Custom',
        'EmailFormatterError',
        `Expected type of 'STR' to be 'string' but received ${getType(str)}`
      );
    }
    return `<em>${str}</em>`;
  },

  /**
   * Formats the given string into strikethrough text.
   * @param {string} str - The text you want to be strikethrough.
   * @returns {string} - The formatted strikethrough text.
   */
  strikethrough(text?: string) {
    const str = text ?? 'Hello, World!';

    if (typeof str !== 'string') {
      throwError(
        'Custom',
        'EmailFormatterError',
        `Expected type of 'STR' to be 'string' but received ${getType(str)}`
      );
    }
    return `<s>${str}</s>`;
  },

  /**
   * Formats the given string with the specified font.
   * @param {Fonts} font - The font name.
   * @param {string} str - The text you want to change the font.
   * @returns {string} - The formatted text with the new font.
   */
  font(font?: Fonts, text?: string) {
    const str = text ?? 'Hello, World!';
    const fontName = font ?? Fonts.SansSerif;
    if (!Object.values(Fonts).includes(fontName)) {
      throwError(
        'Error',
        'InvalidFont',
        `Expected a valid font. But got ${font}`
      );
    }
    if (typeof str !== 'string') {
      throwError(
        'Custom',
        'EmailFormatterError',
        `Expected type of 'STR' to be 'string' but received ${getType(str)}`
      );
    }
    return `<span style="font-family: ${fontName}">${str}</span>`;
  },

  /**
   * Formats the given string into underlined text.
   * @param {string} str - The text you want to be underline.
   * @returns {string} - The formatted underlined text.
   */
  underline(text?: string) {
    const str = text ?? 'Hello, World!';
    if (typeof str !== 'string') {
      throwError(
        'Custom',
        'EmailFormatterError',
        `Expected type of 'STR' to be 'string' but received ${getType(str)}`
      );
    }
    return `<u>${str}</u>`;
  },

  /**
   * Formats the given string with the specified color.
   * @param {string} clr - The color value.
   * @param {string} str - The text you want to change the color.
   * @returns {string} - The formatted text with the new color.
   */
  color(color?: string, text?: string) {
    const str = text ?? 'Hello, World!';
    let clr = color ?? 'black';

    if (typeof clr !== 'string') {
      throwError(
        'Custom',
        'EmailFormatterError',
        `Expected type of 'CLR' to be 'string' but received ${getType(clr)}`
      );
    }
    if (typeof str !== 'string') {
      throwError(
        'Custom',
        'EmailFormatterError',
        `Expected type of 'STR' to be 'string' but received ${getType(str)}`
      );
    }
    return `<span style="color: ${clr}">${str}</span>`;
  },

  /**
   * Formats the given value into a hexadecimal color.
   * @param {string} value - The value to format as hexadecimal.
   * @returns {string} - The formatted hexadecimal color.
   */
  hex(hexString: string) {
    let value = hexString ?? 'FFFFFF';
    if (typeof value !== 'string') {
      throwError(
        'Custom',
        'EmailFormatterError',
        `Expected type of 'value' to be 'string' but received ${getType(value)}`
      );
    }
    return `#${value}`;
  },

  /**
   * Formats the given RGB values into an RGB color.
   * @param {number} r - The red component (0-255).
   * @param {number} g - The green component (0-255).
   * @param {number} b - The blue component (0-255).
   * @returns {string} - The formatted RGB color.
   */
  rgb(red: number, green: number, blue: number) {
    let r = red ?? 255;
    let g = green ?? 0;
    let b = blue ?? 0;

    if (typeof r !== 'number') {
      throwError(
        'Custom',
        'EmailFormatterError',
        `Expected type of 'r' to be 'number' but received ${getType(r)}`
      );
    }
    if (typeof g !== 'number') {
      throwError(
        'Custom',
        'EmailFormatterError',
        `Expected type of 'g' to be 'number' but received ${getType(g)}`
      );
    }
    if (typeof b !== 'number') {
      throwError(
        'Custom',
        'EmailFormatterError',
        `Expected type of 'b' to be 'number' but received ${getType(b)}`
      );
    }
    return `rgb(${r}, ${g}, ${b})`;
  },

  /**
   * Formats the given RGB values and alpha channel into an RGBA color.
   * @param {number} r - The red component (0-255).
   * @param {number} g - The green component (0-255).
   * @param {number} b - The blue component (0-255).
   * @param {number} a - The alpha channel (0-1).
   * @returns {string} - The formatted RGBA color.
   */
  rgba(red: number, green: number, blue: number, alpha: number) {
    let r = red ?? 255;
    let g = green ?? 0;
    let b = blue ?? 0;
    let a = alpha ?? 0.4;

    if (typeof r !== 'number') {
      throwError(
        'Custom',
        'EmailFormatterError',
        `Expected type of 'r' to be 'number' but received ${getType(r)}`
      );
    }
    if (typeof g !== 'number') {
      throwError(
        'Custom',
        'EmailFormatterError',
        `Expected type of 'g' to be 'number' but received ${getType(g)}`
      );
    }
    if (typeof b !== 'number') {
      throwError(
        'Custom',
        'EmailFormatterError',
        `Expected type of 'b' to be 'number' but received ${getType(b)}`
      );
    }
    if (typeof a !== 'number') {
      throwError(
        'Custom',
        'EmailFormatterError',
        `Expected type of 'a' to be 'number' but received ${getType(a)}`
      );
    }
    return `rgba(${r}, ${g}, ${b}, ${a})`;
  },

  /**
   * Formats the given string with the specified font size.
   * @param {string} size - The font size.
   * @param {string} str - The text you want to change the font size.
   * @returns {string} - The formatted text with the new font size.
   */
  size(fontSize?: string, text?: string) {
    const str = text ?? 'Hello, World!';
    let size = fontSize ?? 'medium';

    if (typeof size !== 'string') {
      throwError(
        'Custom',
        'EmailFormatterError',
        `Expected type of 'size' to be 'string' but received ${getType(size)}`
      );
    }
    if (typeof str !== 'string') {
      throwError(
        'Custom',
        'EmailFormatterError',
        `Expected type of 'STR' to be 'string' but received ${getType(str)}`
      );
    }
    return `<span style="font-size: ${size}">${str}</span>`;
  },

  /**
   * Formats the given string with the specified text alignment.
   * @param {Align} align - The text alignment.
   * @param {string} str - The text you want to align.
   * @returns {string} - The formatted text with the new alignment.
   */
  align(textAlign?: Align, text?: string) {
    const align = textAlign ?? Align.left;
    const str = text ?? 'Hello, World!';

    return `<span style="text-align: ${align}">${str}</span>`;
  },
};
